File: greedy/medium/846. Hand of Straights.py
from collections import Counter
from typing import List
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        if len(hand) % groupSize != 0:
            return False
        
        c = Counter(hand)
        hand.sort()
        
        for num in hand:
            if c[num] == 0:
                continue
            for i in range(groupSize):
                if (num + i) not in c or c[num+i] == 0:
                    return False
                c[num+i] -= 1

        return True

File: array/medium/846. Hand of Straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        int n = hand.length;
        if (n % groupSize != 0) return false;

        TreeMap<Integer, Integer> count = new TreeMap<>();

        for (int num : hand) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        while (!count.isEmpty()) {
            int start = count.firstKey();
            for (int i = start; i < start + groupSize; i++) {
                if (count.getOrDefault(i, 0) < 1) return false;
                count.put(i, count.get(i) - 1);
                if (count.get(i) == 0) count.remove(i);
            }
        }

        return true;
    }
}
